ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -0,0 +1,596 @@
+
+    Callback<ReadableStreamChannel> callback = new Callback<ReadableStreamChannel>() {
+      @Override
+      public void onCompletion(final ReadableStreamChannel result, final Exception exception) {
+        if (exception != null) {
+          operationException.set(exception);
+          readCompleteLatch.countDown();
+        } else {
+          Utils.newThread(new Runnable() {
+            @Override
+            public void run() {
+              assertSuccess(result, readCompleteLatch, readCompleteResult, readCompleteException);
+            }
+          }, false).start();
+        }
+      }
+    };
+
+    operationsCount.incrementAndGet();
+    GetBlobOperation op =
+        new GetBlobOperation(routerConfig, routerMetrics, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            callback, operationCompleteCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+      }
+    }
+
+    readCompleteLatch.await();
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    Assert.assertNull("Operation exception should be null, was:" + operationException.get(), operationException.get());
+    Assert.assertNull("Read complete exception should be null", readCompleteException.get());
+    Assert.assertEquals("Size read must equal size written", blobSize, readCompleteResult.get());
+  }
+
+  /**
+   * Assert that the operation is complete and successful. Note that the future completion and callback invocation
+   * happens outside of the GetOperation, so those are not checked here. But at this point, the operation result should
+   * be ready.
+   */
+  private void assertSuccess(ReadableStreamChannel readableStreamChannel, CountDownLatch readCompleteLatch,
+      AtomicLong readCompleteResult, AtomicReference<Exception> readCompleteException) {
+    try {
+      ByteBufferAsyncWritableChannel asyncWritableChannel = new ByteBufferAsyncWritableChannel();
+      long written;
+      Future<Long> readFuture = readableStreamChannel.readInto(asyncWritableChannel, null);
+      // Compare byte by byte.
+      int readBytes = 0;
+      do {
+        ByteBuffer buf = asyncWritableChannel.getNextChunk();
+        while (buf.hasRemaining()) {
+          if (putContent[readBytes++] != buf.get()) {
+            Exception e = new Exception("Content mismatch");
+            readCompleteException.set(e);
+            asyncWritableChannel.resolveOldestChunk(e);
+            return;
+          }
+        }
+        asyncWritableChannel.resolveOldestChunk(null);
+      } while (readBytes < blobSize);
+      written = readFuture.get();
```

Comment:
Yeah same two checks as described in the other comment of mine would be good here I think.